File: pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/java/endpoints/objects_api/memberships/ManageMemberships.java
package com.pubnub.api.java.endpoints.objects_api.memberships;

import com.pubnub.api.java.endpoints.Endpoint;
import com.pubnub.api.java.endpoints.objects_api.utils.Include;
import com.pubnub.api.java.endpoints.objects_api.utils.ObjectsBuilderSteps;
import com.pubnub.api.java.endpoints.objects_api.utils.PNSortKey;
import com.pubnub.api.java.models.consumer.objects_api.membership.PNChannelMembership;
import com.pubnub.api.java.models.consumer.objects_api.membership.PNManageMembershipResult;

import java.util.Collection;

public interface ManageMemberships extends Endpoint<PNManageMembershipResult> {

    ManageMemberships set(Collection<PNChannelMembership> set);

    ManageMemberships remove(Collection<PNChannelMembership> remove);

    ManageMemberships uuid(String uuid);

    ManageMemberships limit(Integer limit);

    ManageMemberships page(com.pubnub.api.models.consumer.objects.PNPage page);

    ManageMemberships filter(String filter);

    ManageMemberships sort(Collection<PNSortKey> sort);

    ManageMemberships includeTotalCount(boolean includeTotalCount);

    ManageMemberships includeCustom(boolean includeCustom);

    ManageMemberships includeChannel(Include.PNChannelDetailsLevel includeChannel);

    interface Builder extends ObjectsBuilderSteps.RemoveOrSetStep<ManageMemberships, PNChannelMembership> {
        @Override
        RemoveStep<ManageMemberships, PNChannelMembership> set(Collection<PNChannelMembership> channelsToSet);

        @Override
        SetStep<ManageMemberships, PNChannelMembership> remove(Collection<PNChannelMembership> channelsToRemove);
    }

}

File: pubnub-kotlin/src/main/kotlin/com/pubnub/api/endpoints/objects/membership/ManageMemberships.kt
package com.pubnub.api.endpoints.objects.membership

import com.pubnub.api.DelegatingEndpoint
import com.pubnub.api.PubNub
import com.pubnub.api.endpoints.remoteaction.ExtendedRemoteAction
import com.pubnub.api.endpoints.remoteaction.MappingRemoteAction
import com.pubnub.api.models.consumer.objects.membership.PNChannelMembershipArrayResult
import com.pubnub.internal.endpoints.objects.membership.IManageMemberships
import com.pubnub.internal.endpoints.objects.membership.ManageMemberships

/**
 * @see [PubNub.manageMemberships]
 */
class ManageMemberships internal constructor(manageMemberships: ManageMemberships) :
    DelegatingEndpoint<PNChannelMembershipArrayResult, com.pubnub.internal.models.consumer.objects.membership.PNChannelMembershipArrayResult>(
        manageMemberships
    ),
    IManageMemberships by manageMemberships {
    override fun convertAction(remoteAction: ExtendedRemoteAction<com.pubnub.internal.models.consumer.objects.membership.PNChannelMembershipArrayResult>): ExtendedRemoteAction<PNChannelMembershipArrayResult> {
        return MappingRemoteAction(remoteAction, PNChannelMembershipArrayResult.Companion::from)
    }
}
